File: D_Zeroing.py
import sys
input = sys.stdin.read().split()
arr_size = int(input[0])
ops = int(input[1])
arr = list(map(int, input[2:]))

arr = [n for n in arr if n != 0]
arr.sort()

prev = 0
count = 0
i = 0

while i < arr_size and count < ops:
    if arr[i] != prev:
        print(arr[i] - prev)
        prev = arr[i]
        count += 1
    i += 1

while count < ops:
    print(0)
    count += 1



# arr_size = int(input[0])
# ops = int(input[1])
# arr = list(map(int, input[2:]))

# arr = sorted(list(set(n for n in arr if n != 0)))

# subs = 0
# count = 0

# for num in arr:
#     if count >= ops:
#         break

#     diff = num - subs
#     print(diff)

#     subs = num
#     count += 1

# while count < ops:
#     print(0)
#     count += 1


# arr_size, ops = map(int, input().split())
# arr = list(map(int, input().split()))
# arr = list(map(int, input().split()))

# for i in range(ops):
#     arr = [n for n in arr if n != 0]

#     if not arr:
#         print(0)
#         continue

#     min_value = min(arr)
#     min_index = arr.index(min(arr))

#     print(min_value)
#     arr = [n - min_value for n in arr]

#     arr[min_index] = 0

    

File: D_Zeroing.cpp
#include <bits/stdc++.h>
using namespace std;

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int arr_size, ops;
    cin >> arr_size >> ops;

    vector<int> arr(arr_size);
    for (int i = 0; i < arr_size; i++){
        cin >> arr[i];
    }

    set<int> unique_arr;
    for (int n : arr){
        if (n != 0){
            unique_arr.insert(n);
        }
    }

    int subs = 0;
    int count = 0;

    vector<int> filtered(unique_arr.begin(), unique_arr.end());
    for (int num : filtered){
        if (count >= ops) break;

        int diff = num - subs;
        cout << diff << "\n";

        subs = num;
        count++;
    }

    while (count < ops){
        cout << 0 << '\n';
        count++;
    }

    return 0;
}